<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Data-Overview" data-toc-modified-id="Data-Overview-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Data Overview</a></span></li><li><span><a href="#Data-Preprocessing" data-toc-modified-id="Data-Preprocessing-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Data Preprocessing</a></span><ul class="toc-item"><li><span><a href="#Handling-Missing-Values" data-toc-modified-id="Handling-Missing-Values-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Handling Missing Values</a></span></li><li><span><a href="#Handling-Anomalous-Values" data-toc-modified-id="Handling-Anomalous-Values-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Handling Anomalous Values</a></span></li><li><span><a href="#Finalizing-Missing-Values" data-toc-modified-id="Finalizing-Missing-Values-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Finalizing Missing Values</a></span></li><li><span><a href="#Changing-Data-Types" data-toc-modified-id="Changing-Data-Types-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Changing Data Types</a></span></li><li><span><a href="#Handling-Duplicates" data-toc-modified-id="Handling-Duplicates-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Handling Duplicates</a></span></li><li><span><a href="#Data-Categorization" data-toc-modified-id="Data-Categorization-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Data Categorization</a></span></li></ul></li><li><span><a href="#Exploratory-Data-Analysis-(EDA)" data-toc-modified-id="Exploratory-Data-Analysis-(EDA)-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Exploratory Data Analysis (EDA)</a></span><ul class="toc-item"><li><span><a href="#Analyze-the-Proportion-of-Debtors-Based-on-Number-of-Children" data-toc-modified-id="Analyze-the-Proportion-of-Debtors-Based-on-Number-of-Children-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Analyze the Proportion of Debtors Based on Number of Children</a></span></li><li><span><a href="#Dependency-Between-Family-Status-and-Loan-Repayment" data-toc-modified-id="Dependency-Between-Family-Status-and-Loan-Repayment-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Dependency Between Family Status and Loan Repayment</a></span></li><li><span><a href="#Dependency-Between-Income-Level-and-Loan-Repayment" data-toc-modified-id="Dependency-Between-Income-Level-and-Loan-Repayment-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Dependency Between Income Level and Loan Repayment</a></span></li><li><span><a href="#Impact-of-Loan-Purpose-on-Repayment" data-toc-modified-id="Impact-of-Loan-Purpose-on-Repayment-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Impact of Loan Purpose on Repayment</a></span></li></ul></li><li><span><a href="#Research-Summary" data-toc-modified-id="Research-Summary-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Research Summary</a></span></li></ul></div>

# Borrower Reliability Assessment: The Impact of Family Status and Number of Children on Loan Repayment

This project was initiated by the bank's credit department to explore the factors that influence a client's ability to repay loans on time. The primary focus is on assessing whether a client's family status, number of children, and other related factors affect their likelihood of defaulting on a loan. The bank provided a dataset containing detailed statistics on clients' payment histories, which serves as the basis for this analysis.

**Objective**

The main objectives of this research are to evaluate the following hypotheses:

- **Family Status:** The marital status of a client impacts the likelihood of timely loan repayment.
- **Number of Children:** The number of children in a family influences the probability of repaying a loan on time.
- **Other Factors:** Additional variables such as income level, loan purpose, and employment type may also play significant roles in repayment behavior.

**Data Overview**

The data for this analysis is sourced from the file `credit_data.csv`, which includes the following key variables:

- `children` — number of children in the family
- `days_employed` — total work experience in days
- `dob_years` — age of the client in years
- `education` — client's education level
- `education_id` — education level identifier
- `family_status` — marital status
- `family_status_id` — marital status identifier
- `gender` — client’s gender
- `income_type` — type of employment
- `debt` — whether the client had debt on loan repayment
- `total_income` — monthly income
- `purpose` — purpose of obtaining the loan

**Research Process**

The study will involve three main stages:

- Data Overview: Initial exploration of the data to understand its structure and content.
- Data Preprocessing: Cleaning and preparing the data for analysis, addressing any critical data quality issues.
- Hypothesis Testing: Conducting statistical analysis to validate or refute the hypotheses.

## Data Overview

**Import the pandas library, then read the data from the CSV file into a DataFrame and save it to the variable `data`.**

In [1]:
import pandas as pd  # Import the pandas library

In [2]:
data = pd.read_csv('/Users/arina/Downloads/my-study-projects/borrower_reliability/credit_data.csv')  # Read the CSV file

**Display the first 20 rows of `data`.**

In [3]:
data.head(20) # Display the first 20 rows of the dataframe data

children  days_employed  dob_years          education  education_id  \
0          1   -8437.673028         42             higher             0   
1          1   -4024.803754         36          secondary             1   
2          0   -5623.422610         33          Secondary             1   
3          3   -4124.747207         32          secondary             1   
4          0  340266.072047         53          secondary             1   
5          0    -926.185831         27             higher             0   
6          0   -2879.202052         43             higher             0   
7          0    -152.779569         50          SECONDARY             1   
8          2   -6929.865299         35             HIGHER             0   
9          0   -2188.756445         41          secondary             1   
10         2   -4171.483647         36             higher             0   
11         0    -792.701887         40          secondary             1   
12         0            NaN         65          secondary             1   
13         0   -1846.641941         54  incomplete higher             2   
14         0   -1844.956182         56             higher             0   
15         1    -972.364419         26          secondary             1   
16         0   -1719.934226         35          secondary             1   
17         0   -2369.999720         33             higher             0   
18         0  400281.136913         53          secondary             1   
19         0  -10038.818549         48          SECONDARY             1   

      family_status  family_status_id gender       income_type  debt  \
0           married                 0      F          employee     0   
1           married                 0      F          employee     0   
2           married                 0      M          employee     0   
3           married                 0      M          employee     0   
4    civil marriage                 1      F           retiree     0   
5    civil marriage                 1      M  business partner     0   
6           married                 0      F  business partner     0   
7           married                 0      M          employee     0   
8    civil marriage                 1      F          employee     0   
9           married                 0      M          employee     0   
10          married                 0      M  business partner     0   
11          married                 0      F          employee     0   
12   civil marriage                 1      M           retiree     0   
13          married                 0      F          employee     0   
14   civil marriage                 1      F  business partner     1   
15          married                 0      F          employee     0   
16          married                 0      F          employee     0   
17   civil marriage                 1      M          employee     0   
18  widower / widow                 2      F           retiree     0   
19         divorced                 3      F          employee     0   

     total_income                             purpose  
0   253875.639453                 purchase of housing  
1   112080.014102                        car purchase  
2   145885.952297                 purchase of housing  
3   267628.550329                additional education  
4   158616.077870                             wedding  
5   255763.565419                 purchase of housing  
6   240525.971920                  housing operations  
7   135823.934197                           education  
8    95856.832424                     for the wedding  
9   144425.938277          buying a home for a family  
10  113943.491460             purchase of real estate  
11   77069.234271  purchase of commercial real estate  
12            NaN                             wedding  
13  130458.228857                        car purchase  
14  165127.911772    purchase of residential property  
15  116820.904450  

**Display basic information about the DataFrame using the `info()` method.**

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


**Conclusion**

The dataset includes 21,525 entries across 12 columns. While most columns are complete, `days_employed` and `total_income` have missing values.

Some columns, like `education` and `purpose`, contain identical categories formatted inconsistently, such as variations in case or word order. This inconsistency will need to be addressed during data preprocessing.

The `days_employed` column, which should represent whole days, is stored as a float64 type and contains negative values, both of which are incorrect and will need correction in the data preprocessing stage.

## Data Preprocessing

### Handling Missing Values

**Display the number of missing values for each column.**

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Two columns have missing values. One of them is `days_employed`, which you will handle in the next step. The other column with missing values is `total_income`, which contains income data. The type of employment has the greatest impact on income, so you should fill in the missing values in this column with the median income value for each type from the `income_type` column. For example, if a person with the employment type `"employee"` has a missing value in the `total_income` column, it should be filled with the median income among all records with the same employment type.**

In [6]:
data['total_income'] = data['total_income'].fillna(data.groupby('income_type')['total_income'].transform("median"))

### Handling Anomalous Values

**The data may contain artifacts (anomalies) — values that do not reflect reality and appeared due to some error. One such artifact is the negative number of days employed in the `days_employed` column. For real data, this is normal. Handle the values in this column by replacing all negative values with positive ones using the `abs()` method.**

In [7]:
data['days_employed'] = data['days_employed'].abs()

**For each type of employment, display the median number of `days employed`.**

In [8]:
data.groupby('income_type')['days_employed'].median()

income_type
business partner        1547.382223
employee                1574.202821
entrepreneur             520.848083
on maternity leave      3296.759962
retiree               365213.306266
state servant           2689.368353
student                  578.751554
unemployed            366413.652744
Name: days_employed, dtype: float64

Two types `unemployed` and `retirees` will have abnormally large values. It is difficult to correct such values, so leave them as they are.

**Display the list of unique values in the `children` column.**

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**There are two anomalous values in the `children` column. Remove the rows containing these anomalous values from `data`.**

In [10]:
data = data[(data['children'] >= 0) & (data['children'] < 20)]

**Display the list of unique values in the `children` column again to ensure that the artifacts have been removed.**

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Finalizing Missing Values

**Fill in the missing values in the `days_employed` column with the median values for each employment type (`income_type`).**

In [12]:
data['days_employed'] = data['days_employed'].fillna(data.groupby('income_type')['days_employed'].transform("median"))

**Ensure that all missing values have been filled. Double-check by displaying the number of missing values for each column using two methods.**

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Changing Data Types

**Convert the data type in the `total_income` column from float to integer using the `astype()` method.**

In [14]:
data['total_income'] = data['total_income'].astype('int')

In [15]:
data['days_employed'] = data['days_employed'].astype('int')

### Handling Duplicates


**Let's check the columns for duplicated values.**

In [16]:
data['education'].unique() 

array(['higher', 'secondary', 'Secondary', 'SECONDARY', 'HIGHER',
       'incomplete higher', 'primary', 'Higher', 'INCOMPLETE HIGHER',
       'Incomplete higher', 'PRIMARY', 'Primary', 'Academic degree',
       'ACADEMIC DEGREE', 'academic degree'], dtype=object)

In [17]:
data['income_type'].unique() 

array(['employee', 'retiree', 'business partner', 'state servant',
       'unemployed', 'entrepreneur', 'student', 'on maternity leave'],
      dtype=object)

In [18]:
data['family_status'].unique() 

array(['married', 'civil marriage', 'widower / widow', 'divorced',
       'single'], dtype=object)

**Handle implicit duplicates in the `education` column. This column contains the same values but recorded differently: with uppercase and lowercase letters. Convert them to lowercase.**

In [19]:
data['education'] = data['education'].str.lower()

**Display the number of duplicate rows in the data. If such rows exist, delete them.**

In [20]:
data[data.duplicated()].head() # Count duplicates

children  days_employed  dob_years  education  education_id  \
1247         0           1572         54  secondary             1   
2849         0           1572         41  secondary             1   
3290         0         365286         58  secondary             1   
4182         1           1572         34     higher             0   
4851         0         365286         60  secondary             1   

       family_status  family_status_id gender income_type  debt  total_income  \
1247  civil marriage                 1      M    employee     0        142594   
2849         married                 0      F    employee     0        142594   
3290  civil marriage                 1      F     retiree     0        118514   
4182  civil marriage                 1      F    employee     0        142594   
4851  civil marriage                 1      F     retiree     0        118514   

                         purpose  
1247                     wedding  
2849  buying a home for a family  
3290                     wedding  
4182                     wedding  
4851                     wedding

In [21]:
data = data.drop_duplicates()   # Remove duplicates

###  Data Categorization

**Based on the ranges indicated below, create a new column in the `data` called `total_income_category` with the following categories:**

- 0–30000 — 'E';
- 30001–50000 — 'D';
- 50001–200000 — 'C';
- 200001–1000000 — 'B';
- 1000001 and above — 'A'.

For example, a borrower with an income of 25000 should be assigned category `'E'`, and a client receiving 235000 — `'B'`. Use a custom function named `categorize_income()` and the `apply()` method.

In [22]:
def categorize_income(total_income):
    if 0 <= total_income <= 30000:
        return 'E'
    if 30001 <= total_income <= 50000:
        return 'D'
    if 50001 <= total_income <= 200000:
        return 'C'
    if 200001 <= total_income <= 1000000:
        return 'B'
    return 'A'

data['total_income_category'] = data['total_income'].apply(categorize_income) 

**Display the list of unique purposes for taking a loan from the `purpose` column.**

In [23]:
data['purpose'].unique()

array(['purchase of housing', 'car purchase', 'additional education',
       'wedding', 'housing operations', 'education', 'for the wedding',
       'buying a home for a family', 'purchase of real estate',
       'purchase of commercial real estate',
       'purchase of residential property', 'construction of own property',
       'real estate', 'construction of real estate',
       'for the purchase of a used car', 'to buy your own car',
       'commercial real estate operations',
       'residential real estate construction', 'housing',
       'operations with your real estate', 'cars', 'take up education',
       'used car deal', 'getting an education', 'car',
       'obtaining additional education', 'buying own housing',
       'real estate operations', 'obtaining higher education', 'your car',
       'car deal', 'specialized education', 'higher education',
       'buying housing for rent', 'for the purchase of a car',
       'housing repairs', 'pursue higher education'], dtype=obj

**Create a function that will create a new column `purpose_category` based on the data in the `purpose` column, with the following categories:**

- `'operations with a car'`,
- `'operations with real estate'`,
- `'wedding planning'`,
- `'getting an education'`.

For example, if the `purpose` column contains the substring `'buying a car'`, then the `purpose_category` column should contain the string `'operations with a car'`.

Use a custom function named `categorize_purpose()` and the `apply()` method. Examine the data in the `purpose` column and determine which substrings will help you correctly categorize it.

In [24]:
def categorize_purpose(purpose):
    # Categories for 'operations with a car'
    if any(keyword in purpose for keyword in [
        'car purchase', 'buying a car', 'buying a used car', 'to buy your own car', 
        'car deal', 'used car deal', 'your car', 'cars', 'car']):
        return 'operations with a car'
    
    # Categories for 'operations with real estate'
    if any(keyword in purpose for keyword in [
        'purchase of housing', 'buying a home for a family', 'purchase of real estate', 
        'purchase of commercial real estate', 'purchase of residential property', 
        'construction of own property', 'real estate', 'construction of real estate',
        'commercial real estate operations', 'residential real estate construction', 
        'operations with your real estate', 'buying own housing', 
        'buying housing for rent', 'housing repairs']):
        return 'operations with real estate'
    
    # Categories for 'wedding planning'
    if any(keyword in purpose for keyword in ['wedding', 'for the wedding', 'to play a wedding']):
        return 'wedding planning'
    
    # Categories for 'getting an education'
    if any(keyword in purpose for keyword in [
        'additional education', 'education', 'getting education', 'take up education', 
        'obtaining additional education', 'specialized education', 
        'higher education', 'obtaining higher education', 'pursue higher education']):
        return 'getting an education'
    
    # Default category for anything that does not fit the above
    return 'other'

# Apply the function
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

## Exploratory Data Analysis (EDA)

### Analyze the Proportion of Debtors Based on Number of Children

**Percentage of Debtors with No Children vs. with Children:**

In [25]:
# Calculate the proportion of debtors with no children
child_zero = data.loc[data['children'] == 0, 'debt']
child_zero_conversion = child_zero.sum() / child_zero.count()
print(f'Proportion of debtors with no children: {child_zero_conversion:.0%}')

# Calculate the proportion of debtors with one or more children
child_more = data.loc[data['children'] > 0, 'debt']
child_more_conversion = child_more.sum() / child_more.count()
print(f'Proportion of debtors with children: {child_more_conversion:.0%}')

# Compare the proportions
if child_zero_conversion > child_more_conversion:
    print('Conclusion: Debtors are more common among people without children.')
else:
    print('Conclusion: Debtors are more common among people with children.')

Proportion of debtors with no children: 8%
Proportion of debtors with children: 9%
Conclusion: Debtors are more common among people with children.


**Percentage of Debtors by Number of Children:**

In [26]:
# Create a pivot table for all data
pivot_table = pd.pivot_table(data, values='debt', index='children', aggfunc=['count', 'sum', 'mean'])
pivot_table.columns = ['total_clients', 'total_debtors', 'debt_ratio']  # Rename columns
pivot_table['debt_ratio'] = (pivot_table['debt_ratio'] * 100).round(2) # Convert debt ratio to percentage

pivot_table

total_clients  total_debtors  debt_ratio
children                                          
0                 14084           1063        7.55
1                  4808            444        9.23
2                  2051            194        9.46
3                   330             27        8.18
4                    41              4        9.76
5                     9              0        0.00

**Conclusion**

Based on the analysis of the most common categories—families with 0, 1, and 2 children—it can be concluded that clients without children are more likely to repay loans on time. As the number of children increases, the proportion of debtors also increases.



### Dependency Between Family Status and Loan Repayment

**Calculate and Print Proportions of Debtors by Family Status**

In [27]:
# Create a pivot table
pivot_table_family = pd.pivot_table(data, values='debt', index='family_status', aggfunc=['count', 'sum', 'mean'])
pivot_table_family.columns = ['total_clients', 'total_debtors', 'debt_ratio'] # Rename columns
pivot_table_family['debt_ratio'] = (pivot_table_family['debt_ratio'] * 100).round(2)  # Convert to percentage

pivot_table_family

total_clients  total_debtors  debt_ratio
family_status                                            
civil marriage            4126            385        9.33
divorced                  1189             84        7.06
married                  12261            927        7.56
single                    2796            273        9.76
widower / widow            951             63        6.62

**Conclusion**

Based on the pivot table, clients who are single are more likely to default on loans compared to those who are married or previously married.

### Dependency Between Income Level and Loan Repayment

**Calculate and Print Proportions of Debtors by Income Category:**

In [28]:
# Create a pivot table
pivot_table_income = pd.pivot_table(data, values='debt', index='total_income_category', aggfunc=['count', 'sum', 'mean'])
pivot_table_income.columns = ['total_clients', 'total_debtors', 'debt_ratio']  # Rename columns
pivot_table_income['debt_ratio'] = (pivot_table_income['debt_ratio'] * 100).round(2)# Convert to percentage

pivot_table_income

total_clients  total_debtors  debt_ratio
total_income_category                                          
A                                 25              2        8.00
B                               5014            354        7.06
C                              15913           1353        8.50
D                                349             21        6.02
E                                 22              2        9.09

**Conclusion**

Based on the comparison of the most common income categories (B and C), it appears that higher income levels are associated with a lower likelihood of loan default.

### Impact of Loan Purpose on Repayment

Calculate and Print Proportions of Debtors by Loan Purpose:

In [29]:
# Create a pivot table
pivot_table_purpose = pd.pivot_table(data, values='debt', index='purpose_category', aggfunc=['count', 'sum', 'mean'])
pivot_table_purpose.columns = ['total_clients', 'total_debtors', 'debt_ratio']  # Rename columns
pivot_table_purpose['debt_ratio'] = (pivot_table_purpose['debt_ratio'] * 100).round(2)# Convert to percentage

pivot_table_purpose

total_clients  total_debtors  debt_ratio
purpose_category                                                     
getting an education                  3988            369        9.25
operations with a car                 4279            400        9.35
operations with real estate           9463            686        7.25
other                                 1288             94        7.30
wedding planning                      2305            183        7.94

**Conclusion**

Clients taking loans for real estate purposes tend to have a lower default rate compared to those taking loans for other purposes.

## Research Summary

The analysis of bank client data reveals several key insights. Factors such as the number of children, marital status, income level, and loan purpose significantly influence the likelihood of timely loan repayment.

- **Children:** Clients without children tend to repay loans more reliably.
- **Marital Status:** Unmarried clients are more likely to default on loans compared to their married counterparts.
- **Income Level:** Although the overall impact of income on loan repayment is not strongly evident, higher-income clients generally show better repayment behavior.
- **Loan Purpose:** Clients who take loans for real estate are more responsible in repayments compared to those borrowing for other purposes.

These findings suggest that the bank should consider these factors when assessing loan risk and making lending decisions to minimize the likelihood of defaults.

**Recommendation:** The bank should gather additional data on clients' income, expenses, and other loans to improve the accuracy of scoring models. Including employment stability, job position, and tenure data could further enhance financial stability assessments.

Implementing these recommendations can improve data quality, scoring accuracy, and reduce the risk of loan defaults.